src/python/pants/backend/native/subsystems/native_toolchain.py

```diff
@@ -63,211 +67,226 @@ def _libc_dev(self):
+    # These mean we don't use any of the headers from our LLVM distribution.
+    '-nobuiltininc',
+  ]
+
   if platform.normalized_os_name == 'darwin':
     xcode_clang = yield Get(CCompiler, XCodeCLITools, native_toolchain._xcode_cli_tools)
-    clang_with_xcode_paths = CCompiler(
+    working_c_compiler = CCompiler(
       path_entries=(provided_clang.path_entries + xcode_clang.path_entries),
       exe_filename=provided_clang.exe_filename,
       library_dirs=(provided_clang.library_dirs + xcode_clang.library_dirs),
-      include_dirs=(xcode_clang.include_dirs + provided_clang.include_dirs))
-    final_llvm_c_compiler = LLVMCCompiler(clang_with_xcode_paths)
+      include_dirs=(provided_clang.include_dirs + xcode_clang.include_dirs),
+      extra_args=(llvm_c_compiler_args + xcode_clang.extra_args))
   else:
     gcc_c_compiler = yield Get(GCCCCompiler, GCC, native_toolchain._gcc)
     provided_gcc = gcc_c_compiler.c_compiler
-    clang_with_gcc_libs = CCompiler(
+    working_c_compiler = CCompiler(
       path_entries=provided_clang.path_entries,
       exe_filename=provided_clang.exe_filename,
-      # We need this version of GLIBCXX to be able to run, unfortunately.
+      # We need g++'s version of the GLIBCXX library to be able to run, unfortunately.
       library_dirs=(provided_gcc.library_dirs + provided_clang.library_dirs),
-      include_dirs=(provided_clang.include_dirs + provided_gcc.include_dirs))
-    final_llvm_c_compiler = LLVMCCompiler(clang_with_gcc_libs)
+      include_dirs=provided_gcc.include_dirs,
+      extra_args=llvm_c_compiler_args)
 
-  yield final_llvm_c_compiler
+  base_linker_wrapper = yield Get(BaseLinker, NativeToolchain, native_toolchain)
+  base_linker = base_linker_wrapper.linker
+  libc_dev = yield Get(LibcDev, NativeToolchain, native_toolchain)
+  working_linker = Linker(
+    path_entries=(base_linker.path_entries + working_c_compiler.path_entries),
+    exe_filename=working_c_compiler.exe_filename,
+    library_dirs=(base_linker.library_dirs + working_c_compiler.library_dirs),
+    linking_library_dirs=libc_dev.get_libc_dirs(platform),
+    extra_args=[])
 
+  yield LLVMCToolchain(llvm_c_compiler=LLVMCCompiler(working_c_compiler),
+                       llvm_c_linker=LLVMCLinker(working_linker))
 
-@rule(LLVMCppCompiler, [Select(Platform), Select(NativeToolchain)])
-def select_llvm_cpp_compiler(platform, native_toolchain):
+
+@rule(LLVMCppToolchain, [Select(Platform), Select(NativeToolchain)])
+def select_llvm_cpp_toolchain(platform, native_toolchain):
   original_llvm_cpp_compiler = yield Get(LLVMCppCompiler, LLVM, native_toolchain._llvm)
-  provided_clangpp = original_llvm_cpp_compiler.cpp_compiler
+  provided_clang = original_llvm_cpp_compiler.cpp_compiler
+
+  # These arguments are shared across platforms.
+  llvm_cpp_compiler_args = [
+    '-x', 'c++', '-std=c++11',
+    # These mean we don't use any of the headers from our LLVM distribution.
+    '-nobuiltininc',
+    '-nostdinc++',
```

Comment:
This implies that we will always be relying on system headers/includes to be there for any calls to `std::*`, correct?

Reply:
No, the opposite -- these flags stop clang from searching for or using the system headers (because we don't control what it finds in that way) except for the ones we provide via `*_INCLUDE_PATH`. We add the system headers that we control / know about specifically later in the `@rule`, either from `XCodeCLITools` or `GCC`.

Reply:
Expanded the definition of these more in 973f26eac. Also, `-nobuiltininc` may not exist, or may not exist on Linux? I'm not sure, but it's not being recognized now -- will see if it's required on OSX.

Reply:
👍